### ISSUE TITLE
Add way to detect and log exceptions in request handlers.

### ISSUE THREAD

**Issue by PatrickNausha** [mswjs/msw#754]
**Is your feature request related to a problem? Please describe.**

I'm currently using `msw` for mocking network requests in `jest` tests. I haven't found an easy way to detect when exceptions are thrown in my request handlers. For example, the code below will produce an exception due to bug and `msw` will silently convert the error to a 500 response.

```js
rest.post("/authorize", (req, res, ctx) => {
	// ... some authorization logic ...
	
	// getCurrentUser throws because there is a bug in it
	const user = getCurrentUser();

	// ... more authorization logic ...

	return res(
		ctx.status(200),
		ctx.json({
			redirectUri,
		}),
	);
});
```

**Describe the solution you'd like**

None of the [lifecycle events](https://mswjs.io/docs/extensions/life-cycle-events#events-list) are emitted when the exception is thrown. A new lifecycle event like `response:error` would allow me to write something like
```js
worker.on('response:error', (res, reqId, error) => {
	console.error("Request handler threw error.", error);
});
```

This would help developers quickly ascertain the source of some test failures without having to follow a chain of symptoms back to a failed `msw` request handler.

**Describe alternatives you've considered**

### Alternative 1
I could use a higher order function wrapper like
```js
rest.post("/authorize", logAndRethrowErrors((req, res, ctx) => {
	// ...
}));
```
```js
function logAndRethrowErrors(fn) {
	return (...args) => {
		try {
			return fn(...args);
		} catch (e) {
			console.error(e);
			throw e;
		}
	};
}
```
Pros:
This code logs an error which likely includes helpful stack trace.

Cons:
This requires all developers on my project to add `logAndRethrowErrors` which may be easy to forget. I'd instead prefer to fail loudly globally.

### Alternative 2
I could do something like this during setup
```js
// Setup
const server = setupServer(...handlers);
const requestsWithoutResponses = new Map();
server.on("request:start", (req) => {
	requestsWithoutResponses.set(req.id, req);
});
server.on("request:end", (req) => {
	requestsWithoutResponses.delete(req.id);
});
```
and this during teardown
```js
// Teardown
server.close();
for (const [, req] of requestsWithoutResponses) {
	console.error(
		"Request received no response. This likely means the request handler threw an error, the error was swallowed by msw, and silently converted to a 500 response.\nRequest:",
		req,
	);
}
```
and assuming any request that did not emit `"request:end"` resulted in an exception.

Pros:
This gives me the global ability fail loudly.

Cons:
Does not have the error message to emit.

### Alternative 3
Combine alternative 1 and 2 above.

Pros:
Allows us to emit error trace with `logAndRethrowErrors` and use the fallback teardown code to hint that a developer forgot `logAndRethrowErrors`.

Cons:
This is quite a few lines of non-standard tooling for something that could be done succinctly if `msw` offered a simple API for it.

**Comment by kettanaito** [comment]
Hey, @PatrickNausha. Thanks for raising this.

Any exceptions occurring in a request handler are gracefully handled as a 500 mocked response. In case your handler fails, your app under test will receive a 500 response instead of a properly mocked response. Here's the responsible logic:

https://github.com/mswjs/msw/blob/8a816e0ff4128c02cae823bc44a3609407641b02/src/utils/worker/createRequestListener.ts#L82-L94

However, this behavior is browser-specific and is not exhibited in Node.js where you're running tests. We should add this behavior to the request processing pipeline to be consistent across different environments. My suggestion would be to integrate it into the `handleRequest` abstraction:

https://github.com/mswjs/msw/blob/8a816e0ff4128c02cae823bc44a3609407641b02/src/node/createSetupServer.ts#L55

**Comment by kettanaito** [comment]
Regarding my comment above, exceptions in the Node's `resolver` function abort the request. It's still up to the request client how it will handle that error. Some clients propagate the exceptions while some handle them in their own way.

I think the request processing pipeline is good as it is, and we should instead consider something as `request:unhandledException` that'd allow developers to react to unhandled exceptions in the handlers.

**Comment by MartinJaskulla** [comment]
1. I would like to work on this, but the issue does not have the "help wanted" tag?
2. What does the tag "needs:tests" mean?

If it is needed in node and the browser:

- Wrap `const result = await handler.run(request, resolutionContext)` in a try catch
- In the catch block `emitter.emit('request:unhandledException', request)`
- Re-throw the error

If it is only needed in node:
- Wrap `return handleRequest<MockedInterceptedResponse>` instead

**Comment by kettanaito** [comment]
Hey, @MartinJaskulla! We'd love to have you involved with this. The issues don't necessarily have to have a "help wanted" label for anybody to join in! We usually use "help wanted" meaning "we can't do this on our own". 

> What does the tag "needs:tests" mean?

It means that this particular issue needs automated tests to be written to verify it. It also means that we don't have such tests, or the tests we have are insufficient in covering it. 

In this particular case, we don't have any tests because the solution to the issue is to extend an existing life-cycle events API with a new `request:unhandledException` event. Tests should be added for that event. 

## What this task is about

First, we need to extend an existing interface of lifecycle events:

https://github.com/mswjs/msw/blob/6fc89f762681aa7f6f697ea78fa5f7c736b35c5d/src/sharedOptions.ts#L17-L24

I propose to add a new `request:unhandledException` event. 

Then we need to go to the common ground between `setupWorker` and `setupServer` that handles request and emit the new event whenever exception occurs during request handler:

https://github.com/mswjs/msw/blob/6fc89f762681aa7f6f697ea78fa5f7c736b35c5d/src/utils/handleRequest.ts#L67-L72

We can wrap `getResponse` in `until` and handle the exception like so:

```ts
const [handlerException, lookupResult] = await until(() => getResponse(...))

if (handlerException) {
  emitter.emit('request:unhandledException', handlerException, handler)

  // Still throw the error to propagate up the request handling pipeline
  // and be transformed into "500 Internal Server Error".
  throw handlerException
}
```

And then cover this scenario with tests. We can use existing tests for reference. I can also help you with tests.

**Comment by MartinJaskulla** [comment]
In your comment you want me to emit the `handler` responsible for the error: `emitter.emit('request:unhandledException', handlerException, handler)`.

However `handlerException` is only the error itself, but it does not contain the responsible handler and lookupResult will be null.

In `getResponse.ts` I would need to wrap this line in a try catch: `const result = await handler.run(request, resolutionContext)` to add the handler to the error to make it available in `handleRequest.ts`.

Alternatively I could emit the request instead of the handler: `emitter.emit('request:unhandledException', request, handlerException)`, which is already available in `handleRequest.ts`.

**Comment by MartinJaskulla** [comment]
What about
- `request:handlerError` instead of `request:unhandledException` (because it is not unhandled)
- Not changing anything in `handleRequest.tsx`
- Wrapping `const result = await handler.run(request, resolutionContext)` in `until` and emitting the error + handler there

Users are only interested in errors occurring inside their handlers, but not all errors that could occur in `getResponse`

**Comment by kettanaito** [comment]
@MartinJaskulla, that sounds like a much better suggestion! I'd still go with `unhandledException` in terms of event name but we can discuss it in more detail later.

**Comment by kettanaito** [comment]
## Released: v0.40.0 🎉

This has been released in v0.40.0!

- 📄 [**Release notes**](https://github.com/mswjs/msw/releases/tag/v0.40.0)
- 📦 [npm package](https://www.npmjs.com/package/msw/v/0.40.0)

Make sure to always update to the latest version (`npm i msw@latest`) to get the newest features and bug fixes.

---

_Predictable release automation by [@ossjs/release](https://github.com/ossjs/release)_.
